src/qililab/qprogram/qblox_compiler.py

```diff
@@ -328,118 +333,137 @@ def _handle_set_offset(self, element: SetOffset):
             else convert(element.offset_path1)
         )
         self._buses[element.bus].qpy_block_stack[-1].append_component(
-            component=QPyInstructions.SetAwgOffs(gain_0=offset_0, gain_1=offset_1)
+            component=QPyInstructions.SetAwgOffs(offset_0=offset_0, offset_1=offset_1)
         )
 
     def _handle_wait(self, element: Wait):
-        convert = QbloxCompiler._convert_value(element)
-        time = (
-            self._buses[element.bus].variable_to_register[element.time]
-            if isinstance(element.time, Variable)
-            else convert(element.time)
+        duration: QPyProgram.Register | int
+        if isinstance(element.duration, Variable):
+            duration = self._buses[element.bus].variable_to_register[element.duration]
+            self._buses[element.bus].dynamic_durations.append(element.duration)
+        else:
+            convert = QbloxCompiler._convert_value(element)
+            duration = convert(element.duration)
+            self._buses[element.bus].static_duration += duration
+        self._buses[element.bus].qpy_block_stack[-1].append_component(
+            component=QPyInstructions.Wait(wait_time=duration)
         )
-        self._buses[element.bus].qpy_block_stack[-1].append_component(component=QPyInstructions.Wait(wait_time=time))
+        self._buses[element.bus].marked_for_sync = True
 
-    def _handle_sync(self, _: Sync):
-        # QBlox does not currently support syncing on selective sequences. If it did we would do somethin like:
-        # buses = element.buses if element.buses is not None else self._buses.keys()
-        for bus in self._buses:
-            self._buses[bus].qpy_block_stack[-1].append_component(
-                component=QPyInstructions.WaitSync(wait_time=QbloxCompiler.minimum_wait_duration)
-            )
+    def _handle_sync(self, element: Sync):
+        # Get the buses involved in the sync operation.
+        buses = set(element.buses or self._buses)
+
+        # If they are zero or one, return.
+        if len(buses) <= 1:
+            return
+
+        # If there is no bus marked for sync, return.
+        if all(not self._buses[bus].marked_for_sync for bus in buses):
+            return
+
+        # Is there any bus that has dynamic durations?
+        if any(bus for bus in buses if self._buses[bus].dynamic_durations or self._buses[bus].sync_durations):
```

Comment:
Isn't the hanlde wait above adding dynamic durations if treated as a variable?
One might want to add dynamic waits for things like 2q gates (eg make the passive qubit wait while the other has a flux applied and iterate for several duration values)

Reply:
Yes, `wait` operation adds a dynamic wait if a variable is used.

Reply:
There are experiments which use duration as a variable. Off the top of my head I can think of the [old cz_conditional experiment](https://github.com/qilimanjaro-tech/HardwareDataTools/blob/bsc-working/src/hwdatatools/experiment_portfolio/cz_conditional.py) but also our current [ssro measurement calibration ](https://github.com/qilimanjaro-tech/LabScripts/blob/main/Personal%20folders/Qililab/ssro_duration.ipynb), they both iterate over duration values for some gate / pulse.
I guess it's not too bad since we could still do this like now through software loops but it'd be very nice if we could do this through hardware loops.
Related to all my other comments about waits / syncs / durations below, I understand it is better to do this in the future / in another pr. So my only concern / doubt (from pure naivety) is if it's gonna be easier to add support for this on top of the current code, or if the current code is such that we're gonna have to modify essential parts of it in order to support duration loops in the future (what do you think?)

Reply:
Yes I know. There is already a task for future PR to support this. No, the changes wouldn't be great, this is why there are separated into different methods since now.